File: src/main/java/zm/unza/counseling/service/interfaces/AppointmentService.java
package zm.unza.counseling.service.interfaces;

import zm.unza.counseling.dto.request.AppointmentRequest;
import zm.unza.counseling.dto.response.AppointmentResponse;
import zm.unza.counseling.dto.AppointmentDto;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

/**
 * Service interface for appointment management
 * Provides abstraction layer for appointment operations
 */
public interface AppointmentService {
    
    /**
     * Create a new appointment
     * 
     * @param request appointment data
     * @return created appointment response
     */
    AppointmentResponse createAppointment(AppointmentRequest request);
    
    /**
     * Update existing appointment
     * 
     * @param id appointment ID
     * @param request updated data
     * @return updated appointment response
     */
    AppointmentResponse updateAppointment(Long id, AppointmentRequest request);
    
    /**
     * Get appointment by ID
     * 
     * @param id appointment ID
     * @return appointment response
     */
    Optional<AppointmentResponse> getAppointmentById(Long id);
    
    /**
     * Get all appointments with pagination
     * 
     * @param page page number
     * @param size page size
     * @return paginated list of appointments
     */
    List<AppointmentResponse> getAllAppointments(int page, int size);
    
    /**
     * Get appointments by client ID
     * 
     * @param clientId client ID
     * @return list of appointments
     */
    List<AppointmentResponse> getAppointmentsByClientId(Long clientId);
    
    /**
     * Get appointments by counselor ID
     * 
     * @param counselorId counselor ID
     * @return list of appointments
     */
    List<AppointmentResponse> getAppointmentsByCounselorId(Long counselorId);
    
    /**
     * Get appointments by date range
     * 
     * @param startDate start date
     * @param endDate end date
     * @return list of appointments
     */
    List<AppointmentResponse> getAppointmentsByDateRange(LocalDateTime startDate, LocalDateTime endDate);
    
    /**
     * Get upcoming appointments
     * 
     * @param limit number of appointments to return
     * @return list of upcoming appointments
     */
    List<AppointmentResponse> getUpcomingAppointments(int limit);
    
    /**
     * Cancel appointment
     * 
     * @param id appointment ID
     * @param reason cancellation reason
     */
    void cancelAppointment(Long id, String reason);
    
    /**
     * Reschedule appointment
     * 
     * @param id appointment ID
     * @param newDateTime new appointment date and time
     * @return updated appointment response
     */
    AppointmentResponse rescheduleAppointment(Long id, LocalDateTime newDateTime);
    
    /**
     * Get appointment statistics
     * 
     * @return appointment statistics
     */
    AppointmentDto getAppointmentStatistics();
    
    /**
     * Check appointment availability
     * 
     * @param counselorId counselor ID
     * @param dateTime appointment date and time
     * @return true if available
     */
    boolean isAppointmentAvailable(Long counselorId, LocalDateTime dateTime);
    
    /**
     * Delete appointment
     * 
     * @param id appointment ID
     */
    void deleteAppointment(Long id);
}

File: src/services/appointmentService.ts
import api from './api';
import { PaginatedResponse } from '../types/api';

/**
 * Appointment Type Definitions
 */
export interface Appointment {
  id: string;
  title: string;
  student: {
    id: string;
    firstName: string;
    lastName: string;
    email: string;
  };
  counselor: {
    id: string;
    firstName: string;
    lastName: string;
    email: string;
  };
  appointmentDate: string;
  duration: number;
  type: 'INITIAL_CONSULTATION' | 'FOLLOW_UP' | 'GROUP_SESSION' | 'ASSESSMENT' | 'CRISIS_INTERVENTION';
  status: 'SCHEDULED' | 'CONFIRMED' | 'IN_PROGRESS' | 'COMPLETED' | 'CANCELLED' | 'NO_SHOW' | 'RESCHEDULED';
  description?: string;
  meetingLink?: string;
  location?: string;
  cancellationReason?: string;
  reminderSent: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface CreateAppointmentRequest {
  studentId: number;
  counselorId: number;
  title: string;
  description?: string;
  appointmentDate: string; // ISO format: "2026-02-05T10:00:00"
  type: 'INITIAL_CONSULTATION' | 'FOLLOW_UP' | 'GROUP_SESSION' | 'ASSESSMENT' | 'CRISIS_INTERVENTION';
}

export interface UpdateAppointmentRequest {
  title?: string;
  description?: string;
  appointmentDate?: string;
  type?: 'INITIAL_CONSULTATION' | 'FOLLOW_UP' | 'GROUP_SESSION' | 'ASSESSMENT' | 'CRISIS_INTERVENTION';
  status?: 'SCHEDULED' | 'CONFIRMED' | 'IN_PROGRESS' | 'COMPLETED' | 'CANCELLED' | 'NO_SHOW' | 'RESCHEDULED';
}

/**
 * Appointment Service - Handles all appointment-related API operations
 * Required Roles: All authenticated users (varies by endpoint)
 */
export const appointmentService = {
  /**
   * Get all appointments (paginated)
   * GET /api/v1/appointments?page=0&size=10
   */
  getAllAppointments: async (page = 0, size = 10): Promise<PaginatedResponse<Appointment>> => {
    const response = await api.get(`/appointments?page=${page}&size=${size}`);
    return response.data.data;
  },

  /**
   * Get appointment by ID
   * GET /api/v1/appointments/{id}
   */
  getAppointmentById: async (id: string): Promise<Appointment> => {
    const response = await api.get(`/appointments/${id}`);
    return response.data.data;
  },

  /**
   * Create new appointment
   * POST /api/v1/appointments
   */
  createAppointment: async (appointmentData: CreateAppointmentRequest): Promise<Appointment> => {
    const response = await api.post('/appointments', appointmentData);
    return response.data.data;
  },

  /**
   * Update appointment
   * PUT /api/v1/appointments/{id}
   */
  updateAppointment: async (id: string, appointmentData: UpdateAppointmentRequest): Promise<Appointment> => {
    const response = await api.put(`/appointments/${id}`, appointmentData);
    return response.data.data;
  },

  /**
   * Get appointments by counselor
   * GET /api/v1/appointments/counselor/{counselorId}
   */
  getAppointmentsByCounselor: async (counselorId: string, page = 0, size = 10): Promise<PaginatedResponse<Appointment>> => {
    const response = await api.get(`/appointments/counselor/${counselorId}?page=${page}&size=${size}`);
    return response.data.data;
  },

  /**
   * Get appointments by student
   * GET /api/v1/appointments/student/{studentId}
   */
  getAppointmentsByStudent: async (studentId: string, page = 0, size = 10): Promise<PaginatedResponse<Appointment>> => {
    const response = await api.get(`/appointments/student/${studentId}?page=${page}&size=${size}`);
    return response.data.data;
  },

  /**
   * Get appointments by client
   * GET /api/v1/appointments/client/{clientId}
   */
  getAppointmentsByClient: async (clientId: string, page = 0, size = 10): Promise<PaginatedResponse<Appointment>> => {
    const response = await api.get(`/appointments/client/${clientId}?page=${page}&size=${size}`);
    return response.data.data;
  },

  /**
   * Get upcoming appointments
   * GET /api/v1/appointments/upcoming
   */
  getUpcomingAppointments: async (page = 0, size = 10): Promise<PaginatedResponse<Appointment>> => {
    const response = await api.get(`/appointments/upcoming?page=${page}&size=${size}`);
    return response.data.data;
  },

  /**
   * Get past appointments
   * GET /api/v1/appointments/past
   */
  getPastAppointments: async (page = 0, size = 10): Promise<PaginatedResponse<Appointment>> => {
    const response = await api.get(`/appointments/past?page=${page}&size=${size}`);
    return response.data.data;
  },

  /**
   * Get cancelled appointments
   * GET /api/v1/appointments/cancelled
   */
  getCancelledAppointments: async (page = 0, size = 10): Promise<PaginatedResponse<Appointment>> => {
    const response = await api.get(`/appointments/cancelled?page=${page}&size=${size}`);
    return response.data.data;
  },

  /**
   * Get confirmed appointments
   * GET /api/v1/appointments/confirmed
   */
  getConfirmedAppointments: async (page = 0, size = 10): Promise<PaginatedResponse<Appointment>> => {
    const response = await api.get(`/appointments/confirmed?page=${page}&size=${size}`);
    return response.data.data;
  },

  /**
   * Get pending appointments
   * GET /api/v1/appointments/pending
   */
  getPendingAppointments: async (page = 0, size = 10): Promise<PaginatedResponse<Appointment>> => {
    const response = await api.get(`/appointments/pending?page=${page}&size=${size}`);
    return response.data.data;
  },

  /**
   * Get today's appointments
   * GET /api/v1/appointments/today
   */
  getTodaysAppointments: async (page = 0, size = 10): Promise<PaginatedResponse<Appointment>> => {
    const response = await api.get(`/appointments/today?page=${page}&size=${size}`);
    return response.data.data;
  },

  /**
   * Cancel appointment
   * PUT /api/v1/appointments/{id}/cancel
   */
  cancelAppointment: async (id: string, reason?: string): Promise<Appointment> => {
    const response = await api.put(`/appointments/${id}/cancel`, { cancellationReason: reason });
    return response.data.data;
  },

  /**
   * Confirm appointment (Admin/Counselor only)
   * PUT /api/v1/appointments/{id}/confirm
   */
  confirmAppointment: async (id: string): Promise<Appointment> => {
    const response = await api.put(`/appointments/${id}/confirm`);
    return response.data.data;
  },

  /**
   * Reschedule appointment
   * PUT /api/v1/appointments/{id}/reschedule
   */
  rescheduleAppointment: async (id: string, newData: {
    appointmentDate: string;
    title?: string;
    description?: string;
  }): Promise<Appointment> => {
    const response = await api.put(`/appointments/${id}/reschedule`, newData);
    return response.data.data;
  },

  /**
   * Check counselor availability
   * GET /api/v1/appointments/availability?counselorId={id}&date={date}
   */
  checkAvailability: async (counselorId: string, date: string): Promise<string[]> => {
    const response = await api.get(`/appointments/availability?counselorId=${counselorId}&date=${date}`);
    return response.data.data;
  },

  /**
   * Get appointment statistics
   * GET /api/v1/appointments/stats
   */
  getAppointmentStatistics: async () => {
    const response = await api.get('/appointments/stats');
    return response.data.data;
  },

  /**
   * Export appointments to CSV/PDF (Admin only)
   * GET /api/v1/appointments/export?format=csv&startDate=&endDate=
   */
  exportAppointments: async (
    format: 'csv' | 'pdf' = 'csv',
    startDate?: string,
    endDate?: string
  ): Promise<Blob> => {
    const params = new URLSearchParams();
    params.append('format', format);
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);

    const response = await api.get(`/appointments/export?${params.toString()}`, {
      responseType: 'blob'
    });
    return response.data;
  }
};
